src/middle/Middle.ml

```diff
@@ -204,6 +210,8 @@ let rec for_scalar st bodyfn var smeta =
   | SVector d | SRowVector d -> mkfor d bodyfn var smeta
   | SMatrix (d1, d2) ->
       mkfor d1 (fun e -> for_scalar (SRowVector d2) bodyfn e smeta) var smeta
+  | SSparseMatrix (d1, _, _, _) ->
+      mkfor d1 bodyfn var smeta
```

Comment:
Here is where I would generate a custom MIR For loop to loop over the size of the nonzero rows / cols.

Reply:
Putting these two comments together because they are on the same topic

> Rather than trying to outsource to the for_scalar function further, you should manually construct a For loop over either the sizedtype dimension of nonzero_rows (or cols) or just over the variable name with a .size() appended; former slightly preferred. Our translation from Stan's "for each" loop does something fairly similar,

I don't really get what you mean by manual here. Here's my understanding.

1. from data_read I have to end up calling `for_scalar_inv` on the size type 
2. `for_scalar_inv` has to return a record that looks like

```
{stmt= For {loopvar; lower; upper; body= {stmt; smeta}}; smeta}
```

i.e. a record with stmt being a statement with a location span and metadata with the other side being an unnamed location span.

Could I just bring `ForEach` in from the ast file, define it in the mir `statements` types where for is, and use that here? I think I'd have to define my first two dimensions (giving the nonzero columns) as some new 1D sized type, but it feels like that would give me what I want?

Reply:
We don't want to add ForEach to the MIR just for this - IR design is a tricky process we want to take seriously. Plus, ForEach is not exactly what you want - you want to iterate over nonzero_rows indices so you can grab both that and the nonzero_cols elements together, right? A ForEach would give you the elements in nonzero_rows with no way to access its corresponding element in nonzero_cols.

But yes, by manually I mean you might have to return that record yourself. It's fine if `upper` is `internal_funapp FnLength [nonzero_rows_var] expr_meta`.